src/pybind/mgr/diskprediction_local/predictor.py

```diff
@@ -39,227 +38,158 @@ class DiskFailurePredictor(object):
+        # change from dict type {'bytes': 123} to just float64 type 123
+        df['user_capacity'] = df['user_capacity'].apply(lambda x: x['bytes'])
 
-        Raises:
-            Exceptions of wrong list/dict operations.
-        """
+        # change from dict type {'table': [{}, {}, {}]}  to list type [{}, {}, {}]
+        df['ata_smart_attributes'] = df['ata_smart_attributes'].apply(lambda x: x['table'])
 
-        all_attrs = [set(disk_day.keys()) for disk_day in disk_days]
-        attr_list = list(set.intersection(*all_attrs))
-        attr_list = disk_days[0].keys()
-        prev_days = disk_days[:-1]
-        curr_days = disk_days[1:]
-        diff_disk_days = []
+        # make a separate column for raw and normalized values of each smart id
+        for day_idx in range(len(disk_days)):
+            for attr_dict in df.iloc[0]['ata_smart_attributes']:
+                smart_id = attr_dict['id']
+                df.at[day_idx, 'smart_{}_raw'.format(smart_id)] = int(attr_dict['raw']['value'])
+                df.at[day_idx, 'smart_{}_normalized'.format(smart_id)] = int(attr_dict['value'])
 
-        for prev, cur in zip(prev_days, curr_days):
-            diff_disk_days.append({attr:(int(cur[attr]) - int(prev[attr]))
-                                   for attr in attr_list})
+        # drop the now-redundant column
+        df = df.drop('ata_smart_attributes', axis=1)
+        return df
 
-        return attr_list, diff_disk_days
 
-    def __get_best_models(self, attr_list):
-        """
-        Find the best model from model list according to given attribute list.
+    def __preprocess(self, disk_days_df):
+        """Scales and transforms input dataframe to feed it to prediction model
 
-        Args:
-            attr_list: All S.M.A.R.T. attributes used in given disk.
+        Arguments:
+            disk_days_df {pandas.DataFrame} -- df where each row holds drive
+                                                features from one day.
 
         Returns:
-            modelpath: The best model for the given attribute list.
-            model_attrlist: 'Ordered' attribute list of the returned model.
-                            Must be aware that SMART attributes is in order.
-
-        Raises:
+            numpy.ndarray -- (n, d) shaped array of n days worth of data and d
+                                features, scaled
         """
-
-        models = self.model_context.keys()
-
-        scores = []
-        for model_name in models:
-            scores.append(sum(attr in attr_list
-                              for attr in self.model_context[model_name]))
-        max_score = max(scores)
-
-        # Skip if too few matched attributes.
-        if max_score < 3:
-            print("Too few matched attributes")
+        # preprocessing may vary across manufactueres. so get manufacturer
+        manufacturer = DiskFailurePredictor.__get_manufacturer(disk_days_df['model_name'].iloc[0]).lower()
```

Comment:
can we just use the "vendor" field? like
```py
manufacturer = disk_days_df['vendor'].iloc[0]
```

see https://github.com/mirror/smartmontools/blob/056603577829363ea38a298a8a2f75690dcb1e52/scsiprint.cpp#L1858-L1862

Reply:
If the smartctl output does include the vendor field then that'd be best. However, I couldn't find the "vendor" field in smartctl data, at least in the Backblaze dataset.
What do you think about doing something like this:
```python3
try:
    manufacturer = raw_df['vendor'].iloc[0]
except KeyError as ke:
    # use log.warn instead of print
    print('"vendor" field not found in smartctl output. Trying to infer from existing data...')
    manufacturer = self.__get_manufacturer(raw_df['model_name'].iloc[0])
```